lib/SILOptimizer/Transforms/PerformanceInliner.cpp

```diff
@@ -505,6 +554,94 @@ static void addWeightCorrection(FullApplySite FAS,
   }
 }
 
+static bool containsWeight(TermInst *inst) {
+  for (auto &succ : inst->getSuccessors()) {
+    if (succ.getCount()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+static void
+addToBBCounts(llvm::DenseMap<SILBasicBlock *, uint64_t> &BBToWeightMap,
+              unsigned long long numToAdd, swift::TermInst *termInst) {
+  for (auto &succ : termInst->getSuccessors()) {
+    auto *currBB = succ.getBB();
+    assert(BBToWeightMap.find(currBB) != BBToWeightMap.end() &&
+           "Expected to find block in map");
+    BBToWeightMap[currBB] += numToAdd;
+  }
+}
+
+static void
+calculateBBWeights(SILFunction *Caller, DominanceInfo *DT,
```

Comment:
This function needs more comments. Most important, a top-level comment which describes what it does.

Also, this looks like a general utility function which might be useful by optimizations other than the inliner. Actually, it would make sense to encapsulate BBToWeightMap into a new utility class and make these functions member functions.